File: java/src/main/java/io/snapvault/model/Commit.java
package io.snapvault.model;

import io.snapvault.hash.Sha256;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.EOFException;
import java.io.IOException;
import java.time.DateTimeException;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/** A snapshot commit linking a root tree to zero or more parent commits. */
public record Commit(String treeId, List<String> parents, Instant timestamp, String message) {
    private static final int MAGIC = 0x53564331; // SVC1
    private static final int MAX_PARENTS = 64;
    private static final int MAX_MESSAGE_BYTES = 4 * 1024 * 1024;

    public Commit {
        Sha256.requireObjectId(Objects.requireNonNull(treeId, "treeId"));
        parents = List.copyOf(Objects.requireNonNull(parents, "parents"));
        if (parents.size() > MAX_PARENTS) {
            throw new IllegalArgumentException("A commit cannot have more than 64 parents");
        }
        for (String parent : parents) {
            Sha256.requireObjectId(parent);
        }
        Objects.requireNonNull(timestamp, "timestamp");
        message = Objects.requireNonNull(message, "message");
        if (message.indexOf('\0') >= 0) {
            throw new IllegalArgumentException("Commit message cannot contain NUL");
        }
        if (message.getBytes(java.nio.charset.StandardCharsets.UTF_8).length > MAX_MESSAGE_BYTES) {
            throw new IllegalArgumentException("Commit message is too large");
        }
    }

    public byte[] encode() throws IOException {
        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
        try (DataOutputStream output = new DataOutputStream(bytes)) {
            output.writeInt(MAGIC);
            output.write(Sha256.bytes(treeId));
            output.writeInt(parents.size());
            for (String parent : parents) {
                output.write(Sha256.bytes(parent));
            }
            output.writeLong(timestamp.getEpochSecond());
            output.writeInt(timestamp.getNano());
            Tree.writeString(output, message);
        }
        return bytes.toByteArray();
    }

    public static Commit decode(byte[] encoded) throws IOException {
        Objects.requireNonNull(encoded, "encoded");
        try (ByteArrayInputStream bytes = new ByteArrayInputStream(encoded);
                DataInputStream input = new DataInputStream(bytes)) {
            if (input.readInt() != MAGIC) {
                throw new IOException("Invalid commit object signature");
            }

            String treeId = readObjectId(input, "tree");
            int parentCount = input.readInt();
            if (parentCount < 0 || parentCount > MAX_PARENTS) {
                throw new IOException("Invalid commit parent count: " + parentCount);
            }
            List<String> parents = new ArrayList<>(parentCount);
            for (int index = 0; index < parentCount; index++) {
                parents.add(readObjectId(input, "parent"));
            }

            Instant timestamp;
            try {
                long epochSecond = input.readLong();
                int nanosecond = input.readInt();
                if (nanosecond < 0 || nanosecond > 999_999_999) {
                    throw new IOException("Invalid commit nanosecond: " + nanosecond);
                }
                timestamp = Instant.ofEpochSecond(epochSecond, nanosecond);
            } catch (DateTimeException exception) {
                throw new IOException("Invalid commit timestamp", exception);
            }
            String message = Tree.readString(input, MAX_MESSAGE_BYTES);
            if (bytes.available() != 0) {
                throw new IOException("Commit object has trailing data");
            }

            try {
                return new Commit(treeId, parents, timestamp, message);
            } catch (IllegalArgumentException exception) {
                throw new IOException("Invalid commit object", exception);
            }
        } catch (EOFException exception) {
            throw new IOException("Truncated commit object", exception);
        }
    }

    private static String readObjectId(DataInputStream input, String description) throws IOException {
        byte[] objectId = input.readNBytes(Sha256.BYTE_LENGTH);
        if (objectId.length != Sha256.BYTE_LENGTH) {
            throw new EOFException("Truncated " + description + " object id");
        }
        return Sha256.hex(objectId);
    }
}

File: go/internal/object/commit.go
package object

import (
	"errors"
	"fmt"
	"slices"
	"strings"
	"time"
)

const (
	commitMagic     = 0x53564331 // "SVC1"
	maxParents      = 64
	maxMessageBytes = 4 << 20
)

// Commit is a snapshot commit linking a root tree to zero or more parents.
type Commit struct {
	TreeID  string
	Parents []string
	Time    time.Time
	Message string
}

// NewCommit validates and builds a commit.
func NewCommit(treeID string, parents []string, at time.Time, message string) (*Commit, error) {
	if err := RequireID(treeID); err != nil {
		return nil, err
	}
	if len(parents) > maxParents {
		return nil, errors.New("a commit cannot have more than 64 parents")
	}
	for _, parent := range parents {
		if err := RequireID(parent); err != nil {
			return nil, err
		}
	}
	if strings.ContainsRune(message, 0) {
		return nil, errors.New("commit message cannot contain NUL")
	}
	if len(message) > maxMessageBytes {
		return nil, errors.New("commit message is too large")
	}
	return &Commit{
		TreeID:  treeID,
		Parents: slices.Clone(parents),
		Time:    at,
		Message: message,
	}, nil
}

// Encode returns the canonical binary payload defined by format v1.
func (c *Commit) Encode() []byte {
	encoded := appendInt32(nil, commitMagic)
	encoded = appendIDBytes(encoded, c.TreeID)
	encoded = appendInt32(encoded, int32(len(c.Parents)))
	for _, parent := range c.Parents {
		encoded = appendIDBytes(encoded, parent)
	}
	encoded = appendInt64(encoded, c.Time.Unix())
	encoded = appendInt32(encoded, int32(c.Time.Nanosecond()))
	return appendUTF8String(encoded, c.Message)
}

func appendIDBytes(dst []byte, id string) []byte {
	raw, err := idBytes(id)
	if err != nil {
		panic(fmt.Sprintf("commit holds an invalid object id: %v", err))
	}
	return append(dst, raw...)
}

// DecodeCommit parses and validates a commit payload.
func DecodeCommit(payload []byte) (*Commit, error) {
	commit, err := decodeCommit(&payloadReader{rest: payload})
	if errors.Is(err, errTruncated) {
		return nil, errors.New("truncated commit object")
	}
	return commit, err
}

func decodeCommit(r *payloadReader) (*Commit, error) {
	magic, err := r.int32()
	if err != nil {
		return nil, err
	}
	if magic != commitMagic {
		return nil, errors.New("invalid commit object signature")
	}
	treeID, err := r.objectID()
	if err != nil {
		return nil, err
	}
	parentCount, err := r.int32()
	if err != nil {
		return nil, err
	}
	if parentCount < 0 || parentCount > maxParents {
		return nil, fmt.Errorf("invalid commit parent count: %d", parentCount)
	}
	parents := make([]string, 0, parentCount)
	for range parentCount {
		parent, err := r.objectID()
		if err != nil {
			return nil, err
		}
		parents = append(parents, parent)
	}
	seconds, err := r.int64()
	if err != nil {
		return nil, err
	}
	nanos, err := r.int32()
	if err != nil {
		return nil, err
	}
	if nanos < 0 || nanos > 999_999_999 {
		return nil, fmt.Errorf("invalid commit nanosecond: %d", nanos)
	}
	message, err := r.utf8String(maxMessageBytes)
	if err != nil {
		return nil, err
	}
	if r.remaining() != 0 {
		return nil, errors.New("commit object has trailing data")
	}

	commit, err := NewCommit(treeID, parents, time.Unix(seconds, int64(nanos)), message)
	if err != nil {
		return nil, fmt.Errorf("invalid commit object: %w", err)
	}
	return commit, nil
}
